modules/mqtt/destination/mqtt-destination.c

```diff
@@ -0,0 +1,280 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "mqtt-destination.h"
+#include "mqtt-worker.h"
+#include "mqtt-parser.h"
+
+#include "plugin.h"
+#include "messages.h"
+#include "misc.h"
+#include "stats/stats-registry.h"
+#include "logqueue.h"
+#include "driver.h"
+#include "plugin-types.h"
+#include "logthrdest/logthrdestdrv.h"
+
+/*
+ * Default values
+ */
+
+#define DEFAULT_HOST "localhost"
+#define DEFAULT_PORT 1833
+#define DEFAULT_CLEAN_SESSION FALSE
+#define DEFAULT_KEEPALIVE 60
+#define DEFAULT_QOS 0
+#define DEFAULT_PROTOCOL_TYPE "tcp"
+#define DEFAULT_MESSAGE_TEMPLATE "$MESSAGE"
+
+
+/*
+ * Configuration
+ */
+
+void
+mqtt_destination_dd_set_host (LogDriver *d, const gchar *host)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  g_string_assign(self->host, host);
+}
+
+void
+mqtt_destination_dd_set_port (LogDriver *d, const gint port)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  self->port = port;
+}
+
+void
+mqtt_destination_dd_set_topic(LogDriver *d, const gchar *topic)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  g_string_assign(self->topic, topic);
+}
+
+
+void
+mqtt_destination_dd_set_keepalive (LogDriver *d, const gint keepalive)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  self->keepalive = keepalive;
+}
+
+void
+mqtt_destination_dd_set_qos (LogDriver *d, const gint qos)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  self->qos = qos;
+}
+
+void
+mqtt_destination_dd_set_protocol_type(LogDriver *d, const gchar *protocol_type)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  g_string_assign(self->protocol_type, protocol_type);
+}
+
+void
+mqtt_destination_dd_set_message_ref(LogDriver *d, LogTemplate *message)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+
+  log_template_unref(self->message);
+  self->message = message;
+}
+
+/*
+ * Utilities
+ */
+static const gchar *
+_format_stats_instance(LogThreadedDestDriver *d)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  static gchar stats_instance[1024];
+
+  if (((LogPipe *)d)->persist_name)
+    g_snprintf(stats_instance, sizeof(stats_instance), "%s", ((LogPipe *)d)->persist_name);
+  else
+    g_snprintf(stats_instance, sizeof(stats_instance),
+              "mqtt-destination,%s,%d,%s", self->host->str, self->port, self->topic->str);
```

Comment:
the stats framework has a specific field for annotating a counter as "source" or "destination". the name itself shouldn't contain it (and neither does kafka/file/socket for instance).

Reply:
solved by this commit: MQTT/destination: remove destination prefix from stats_instance
(commit hash: eb15dedf404637b7cfc11504784fda692a90f899 )